### PR TITLE
Graphite: Support request cancellation properly (Uses new backendSrv.fetch Observable request API)

### PR DESCRIPTION
**Pull request by mckn**
**What this PR does / why we need it**:
This PR replaces the usage of datasourceRequest in the Graphite data source with fetch and updates test accordingly.

**Which issue(s) this PR fixes**:
Relates #27222

**Special notes for your reviewer**:

### CODE REVIEW

File: public/app/plugins/datasource/graphite/datasource.test.ts [hunk 1]
```diff
@@ -128,9 +130,12 @@ describe('graphiteDatasource', () => {
...
 
-      await ctx.ds.query(query as any).then((data: any) => {
-        results = data;
-      });
+      await ctx.ds
+        .query(query as any)
+        .toPromise()
+        .then((data: any) => {
+          results = data;
+        });
```
Comment (hugohaggmark) [inline public/app/plugins/datasource/graphite/datasource.test.ts:138]: This looks a bit weird with an `await` and a Promise chain? And shouldn't we test the Observable not the Promise?

File: public/app/plugins/datasource/graphite/datasource.test.ts [hunk 2]
```diff
@@ -197,8 +202,8 @@ describe('graphiteDatasource', () => {
       };
 
       beforeEach(async () => {
-        datasourceRequestMock.mockImplementation((options: any) => {
-          return Promise.resolve(response);
+        fetchMock.mockImplementation((options: any) => {
+          return createFetchResponse(response);
```
Comment (hugohaggmark) [inline public/app/plugins/datasource/graphite/datasource.test.ts:206]: nit: I see a lot of mockImplementation() but I'm missing a mockClear() somewhere. Maybe I need to look harder.

File: public/app/plugins/datasource/graphite/datasource.test.ts [hunk 3]
```diff
@@ -503,3 +508,22 @@ accessScenario('with direct access', 'http://localhost:8080', (httpOptions: any)
   expect(httpOptions.headers['X-Dashboard-Id']).toBe(undefined);
   expect(httpOptions.headers['X-Panel-Id']).toBe(undefined);
 });
+
+const createFetchResponse = (response: Partial<FetchResponse>): Observable<FetchResponse> => {
```
Comment (hugohaggmark) [inline public/app/plugins/datasource/graphite/datasource.test.ts:512]: I think we already have a shared function for this in `public/test/helpers/createFetchResponse.ts`.

File: public/app/plugins/datasource/graphite/datasource.ts [hunk 4]
```diff
@@ -473,15 +491,19 @@ export class GraphiteDatasource extends DataSourceApi<GraphiteQuery, GraphiteOpt
...
+      .pipe(
+        map((results: any) => {
+          if (results.data) {
+            return _.map(results.data, (tag) => {
+              return { text: tag };
+            });
+          } else {
+            return [];
+          }
+        })
```
Comment (hugohaggmark) [inline public/app/plugins/datasource/graphite/datasource.ts:504]: This code is shared between `getTagsAutoComplete` and `getTagValuesAutoComplete`, maybe we can create a `OperatorFunction` both can share?

**Review by hugohaggmark — COMMENTED** [review]
Looking good and I left some comments for consideration 🎉

**Comment by hugohaggmark** [comment]
Also, I'm not sure we should push changes like these into 7.5.0? If we still consider adding this to 7.5.0 then we should also add the `backport v7.5.x` label.

**Comment by mckn** [comment]
@hugohaggmark awesome feedback! I will have a look at it during the day 🙌

File: public/app/plugins/datasource/graphite/datasource.test.ts (hunk 2 above)
Comment (mckn) [reply public/app/plugins/datasource/graphite/datasource.test.ts:206]: You have a `jest.clearAllMocks();` in the top `beforeEach` that should cover this.

Comment (hugohaggmark) [reply public/app/plugins/datasource/graphite/datasource.test.ts:206]: Oops, I missed that 🙈

**Review by dprokop — APPROVED** [review]
Looks good! Agree with Hugo's comments:)

File: public/app/plugins/datasource/graphite/datasource.ts [hunk 5]
```diff
@@ -699,3 +704,17 @@ function supportsTags(version: string): boolean {
 function supportsFunctionIndex(version: string): boolean {
   return isVersionGtOrEq(version, '1.1');
 }
+
+function mapToTags(): OperatorFunction<any, Array<{ text: string }>> {
```
Comment (mckn) [inline public/app/plugins/datasource/graphite/datasource.ts:708]: @hugohaggmark @torkelo do you know if we have any interface or types for graphite tags?

Comment (hugohaggmark) [reply public/app/plugins/datasource/graphite/datasource.ts:708]: Not sure

**Review by hugohaggmark — APPROVED** [review]
LGTM!

**Pull request merged**